aikau/src/main/resources/alfresco/lists/Paginator.js

```diff
@@ -755,6 +767,99 @@ define(["dojo/_base/declare",
+
+          // only need to act if not initialized yet or change occurred
+          firstPage = 0;
+          lastPreloadedPage = -1;
+          
+          // Delete the (obsolete) previous page selector group contents...
+          array.forEach(this.pageSelectorGroup.getChildren(), function(widget, index, arr) {
+              var virtPageEnd;
+              
+              // need to destroy every existing menu item if documentsPerPage has changed
+              // or state hasn't been fully initialized yet
+              if (this._pageSelectionDocsPerPage === undefined
+                      || this._pageSelectionDocsPerPage !== this.documentsPerPage
+                      || this._pageSelectionTotalRecords === undefined)
+              {
+                  this.pageSelectorGroup.removeChild(widget);
+                  widget.destroy();
+              }
+              // otherwise if totalRecords changed we need to destroy last page and any obsolete page before
+              // calculate pageEnd for existing page and compare
+              else if (this._pageSelectionTotalRecords !== this.totalRecords)
+              {
+                  virtPageEnd = (index + 1) * parseInt(this.documentsPerPage, 10);
+                  if (index === arr.length - 1 || virtPageEnd > this.totalRecords)
+                  {
+                      this.pageSelectorGroup.removeChild(widget);
+                      widget.destroy();
+                      
+                      // mark the first page that needs re-creation
+                      if (firstPage === 0)
+                      {
+                         firstPage = index; 
+                      }
+                  }
+                  else
+                  {
+                      lastPreloadedPage = index;
+                  }
+              }
+              else
+              {
+                  lastPreloadedPage = index;
+              }
+           }, this);
+
+          // Create the page labels, which for English will be along the lines of 1-25
+          firstPage = firstPage || (lastPreloadedPage + 1);
+          pageStart = firstPage * parseInt(this.documentsPerPage, 10) + 1;
+          for (i = firstPage; i < this.totalPages; i++)
+          {
+             // Comments below assume 25 docs per page...
+             if (i+1 !== this.totalPages)
+             {
+                // If we're not getting the labels for the last page...
+                pageEnd = pageStart + parseInt(this.documentsPerPage, 10) - 1; // Deduct 1 because it's 1 - 25 (not 1 - 26!)
+             }
+             else
+             {
+                // ...for the last page just count up to the last document
+                pageEnd = this.totalRecords;
+             }
+
+             label = this.message("list.paginator.page.label", {0: pageStart, 1: pageEnd, 2: this.totalRecords});
+             menuItem = new AlfCheckableMenuItem({
```

Comment:
That's good to know. I've always disliked the new-ing as it often required quite nasty overrides/extensions to adapt when necessary.
I hope the Surf dependency analysis regexes will be enhanced (or overriden by Aikau module) so we don't have to keep adding the widget module IDs used in the code (call to createWidget) to the define-call (also always hated to have been forced to add modules there just for the dependency analysis).